hub/api/tensor.py

```diff
@@ -25,29 +33,49 @@ def __init__(
             key (str): The internal identifier for this tensor.
             provider (StorageProvider): The storage provider for the parent dataset.
             tensor_slice (slice): The slice object restricting the view of this tensor.
+            tensor_meta (dict): For internal use only. If a tensor with `key` doesn't exist, a new tensor is created with this meta.
+
+        Raises:
+            TensorDoesNotExistError: If no tensor with `key` exists and a `tensor_meta` was not provided.
         """
         self.key = key
         self.provider = provider
         self.slice = tensor_slice
 
-        self.load_meta()
+        if not tensor_exists(self.key, self.provider):
+            if tensor_meta is None:
+                raise TensorDoesNotExistError(self.key)
+
+            create_tensor(self.key, self.provider, tensor_meta)
 
-    def load_meta(self):
-        meta = read_tensor_meta(self.key, self.provider)
-        self.num_samples = meta["length"]
-        self.shape = meta["max_shape"]
+    def append(self, array: np.ndarray, batched: bool):
+        add_samples_to_tensor(
+            array,
+            self.key,
+            storage=self.provider,
+            batched=batched,
+        )
+
+    @property
+    def meta(self):
+        return read_tensor_meta(self.key, self.provider)
+
+    @property
+    def shape(self):
+        # TODO: when dynamic arrays are supported, handle `min_shape != max_shape` (right now they're always equal)
+        return self.meta["max_shape"]
 
     def __len__(self):
-        """Return the length of the primary axis"""
-        return self.num_samples
+        """Return the length of the primary axis."""
+        return self.meta["length"]
 
     def __getitem__(self, item: Union[int, slice]):
         if isinstance(item, int):
             item = slice(item, item + 1)
 
         if isinstance(item, slice):
             new_slice = merge_slices(self.slice, item)
-            return Tensor(self.key, self.provider, new_slice)
+            return Tensor(self.key, self.provider, tensor_slice=new_slice)
```

Comment:
The `Index` PR got merged, right? I think you might need to merge the upstream branch